design/milestone_1_2_design_doc.md

```diff
@@ -0,0 +1,430 @@
+        * [Delete](#delete)
+    + [Order of Deployment](#order-of-deployment)
+      - [Milestone 1](#milestone-1-1)
+    + [Backwards compatibility](#backwards-compatibility)
+      - [Milestone 1](#milestone-1-2)
+    + [Performance](#performance)
+      - [Milestone 1](#milestone-1-3)
+    + [Affected Components](#affected-components)
+  * [Security](#security)
+  * [Test Plan](#test-plan)
+    + [Integration tests](#integration-tests)
+    + [Performance tests](#performance-tests)
+  * [Logs](#logs)
+  * [Documentation](#documentation)
+  * [Open questions](#open-questions)
+  * [Implementation plan](#implementation-plan)
+
+## Glossary
+
+| **Term**                                                     | **Description**                                              |
+| ------------------------------------------------------------ | ------------------------------------------------------------ |
+| [Job](*https://kubernetes.io/docs/concepts/workloads/controllers/job/*) | K8s entity that runs a pod for one-time operation. Once the pod exists, the Job terminates |
+| [Deployment](*https://kubernetes.io/docs/concepts/workloads/controllers/deployment/*) | K8s entity that ensures multiple replications of a pod are running all the time. If a pod terminates, another one will be created. |
+| [Kubernetes Helm](https://helm.sh/)                          | A tool that streamlines and organizes the management of Kubernetes installation, deployment, and upgrade processes. |
+| [Helm Chart](https://helm.sh/docs/topics/charts/)            | A collection of files that describe a related set of Kubernetes resources. A chart can be used to describe a simple pod of a complex application. |
+
+## Useful links
+
+| Name                | Link                                                         |
+| ------------------- | ------------------------------------------------------------ |
+| Phase 2 Feature Doc | https://github.com/cyberark/secrets-provider-for-k8s/issues/102 |
+
+## Background
+
+The Cyberark Secrets Provider for K8s enables customers to use secrets stored and managed in Conjur/DAP vault and consume them as K8s Secrets for their application containers.
+
+During Phase 1, an *init container* was deployed in the same Pod as the customer's applications to populate the K8s secrets defined in the pod application's manifest.
+
+### Motivation
+
+The current implementation of the Secrets Provider only supports one application and makes the deployment of the Secrets Provider deeply coupled to the customer's applications. 
+
+## Requirements
+
+The solution should support the following capabilities:
+
+###  Milestone 1 *(current)*
+
+- Secrets Provider runs as a separate entity, serving multiple application containers that run on multiple pods
+- Solution needs to be native to Kubernetes
+- Lifecycle should support update/removal of deployment
+- Provide a way for customer to understand the state of the Secret Provider - when it finished initializing
```

Comment:
I think we should provide more details on how will the customer understand the state of secret provider and what are the features we should develop for it

Reply:
I added that we will supply a log detailing if it is a job or a deployment
> Kubernetes Secrets Provider v\*%s\* starting up as a %s... (Job/Deployment)

Is that enough in your opinion?

Reply:
I think we should add some indication that secret provider is stuck - maybe using liveliness probe.
Also - we should add a printout saying when the secrets where lately copied from conjur, as well as error message with secrets whose retrieval / copy failed

Reply:
Failures in secret retrieval already exist. For example:

```
const CSPFK020E string = "CSPFK020E Failed to retrieve k8s secret"
const CSPFK021E string = "CSPFK021E Failed to retrieve k8s secrets"
const CSPFK022E string = "CSPFK022E Failed to update k8s secret"
const CSPFK023E string = "CSPFK023E Failed to update K8s secrets"
```
These log entries are timestamped

in terms of liveliness probe under what types of circumstances would the secrets provider be stuck?

Reply:
it can be bug, out-of-memory on computer, network issue or any other problem

Reply:
This seems like an addition in Milestone 2 when we are deployed as a deployment and not a job but this is a good point. 

From what I see [here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/): 

A ping must be sent to /tmp/healthy or /healthz (using HTTP) of the Secrets Provider and expect a 0 or return 200 response which I can imagine we can wrap that in code. Regarding `periodSeconds` for how often the Secrets Provider should be probed, how often is conventional?

`initialDelaySeconds` for how long the initial probe should wait before performing the first probe, what is conventional? 5 or 10 seconds?

We can also via HTTP request but why after X seconds (in the link HTTP example, 10 seconds) an unhealthy response is returned? why is there a limit for how long the container is deemed healthy?

Reply:
Sigal - what you wrote looks a right direction. 
regarding periodSeconds -the typical value is 3 seconds
initialDelaySeconds - you can try 5 seconds
There is a limit of time after which if container does not return a response it is considered to be stuck - exactly the same way as you consider the regular application stuck if it does not respond